[CALCITE-6824] JOIN_SUB_QUERY_TO_CORRELATE rule produces an incorrect…

suibianwanwank opened the pull request:
… plan when operands are not empty

suibianwanwank commented on core/src/test/resources/sql/sub-query.iq:
```diff
@@ -405,6 +405,78 @@ EnumerableCalc(expr#0..4=[{inputs}], EMPNO=[$t0])
       EnumerableTableScan(table=[[scott, DEPT]])
 !plan
 
+# [CALCITE-6824] JOIN_SUB_QUERY_TO_CORRELATE rule produces an incorrect plan when operands are not empty
```
@asolimando I used the test given by Julian here. If required I can add it to JdbcTest as well.

asolimando reviewed (changes requested):
No major remarks at this stage but I have requested some changes and few suggestions. Please check also the comments I left in Jira as it will affect the title and the associated commit message/comments in the code.

suibianwanwank commented:
@asolimando Thank you for your detailed review! I have addressed and fixed the issues you mentioned. If I missed anything or made any mistakes, please let me know.

asolimando commented on core/src/main/java/org/apache/calcite/rel/rules/SubQueryRemoveRule.java:
```diff
@@ -915,43 +915,50 @@ private static void matchJoin(SubQueryRemoveRule rule, RelOptRuleCall call) {
-    if (inputSet.intersects(ImmutableBitSet.range(0, nFieldsLeft))
-        && inputSet.intersects(ImmutableBitSet.range(nFieldsLeft, nFieldsLeft + nFieldsRight))) {
+
+    ImmutableBitSet inputIntersectsLeftSide = ImmutableBitSet.range(0, nFieldsLeft);
```
What is meant is to save the boolean value directly in a final variable, if you prefer to keep this form let's at least adapt the name of the variables to `leftInputFields`/`rightInputFields` and make them final.

asolimando commented on core/src/main/java/org/apache/calcite/rel/rules/SubQueryRemoveRule.java:
```diff
@@ -915,43 +915,50 @@ private static void matchJoin(SubQueryRemoveRule rule, RelOptRuleCall call) {
     }
 
-    if (inputSet.intersects(ImmutableBitSet.range(0, nFieldsLeft))) {
+    final Set<CorrelationId>  variablesSet = RelOptUtil.getVariablesUsed(e.rel);
```
```suggestion
    final Set<CorrelationId> variablesSet = RelOptUtil.getVariablesUsed(e.rel);
```
Nit: double space

asolimando commented on core/src/main/java/org/apache/calcite/rel/rules/SubQueryRemoveRule.java:
```diff
@@ -915,43 +915,50 @@ private static void matchJoin(SubQueryRemoveRule rule, RelOptRuleCall call) {
 
       final int nFields = builder.fields().size();
       ImmutableList<RexNode> fields =
-          builder.fields(ImmutableBitSet.range(0, nFieldsLeft)
```
Now I see why you are saving the `ImmutableBitSet`, but I think it's better to save the boolean value of the intersect and rebuild the range here (so basically reverting this part of the change).

asolimando commented on core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java:
```diff
@@ -9119,9 +9119,9 @@ public interface Config extends RelRule.Config {
-  @Test void testJoinSubQueryRemoveRule() {
+   * Subquery in join conditions rewrite fails if referencing a column
+   * from the right-hand side table</a>. */
+  @Test void testJoinSubQueryRemoveRuleWithQuantifierNotIn() {
```
```suggestion
  @Test void testJoinSubQueryRemoveRuleWithNotIn() {
```

asolimando commented on core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml:
```diff
@@ -6285,7 +6285,7 @@ LogicalProject(SAL=[$5])
     </Resource>
   </TestCase>
-  <TestCase name="testJoinSubQueryRemoveRule">
+  <TestCase name="testJoinSubQueryRemoveRuleWithQuantifierNotIn">
```
```suggestion
  <TestCase name="testJoinSubQueryRemoveRuleWithNotIn">
```

asolimando reviewed (approved):
Thanks @suibianwanwank for applying the changes.

I have left few minor change requests, after that I feel this is ready to be merged.

suibianwanwank commented:
@asolimando Done.
